File: Algorithms/counting sort/counting_sort.c
#include <stdio.h>
#include <stdlib.h>









int max(int arr[], int n)
{
    int max = arr[0];
    for (int i = 1; i < n; i++)
        if (arr[i] > max)
            max = arr[i];

    return max;
}





void counting_sort(int *array, int n, int k) 
{
    int count[k+1];
    for (int i = 0; i < k+1; i++)
        count[i] = 0;

    for (int j = 0; j < n; j++)
        count[array[j]]++;

    for (int i = 1; i < k+1; i++)
        count[i] += count[i-1];


    int out[n];
    for (int i = n-1; i >= 0; i--)
    {
        out[count[array[i]]-1] = array[i];
        count[array[i]]--;
    }

    for (int j = 0; j < n; j++)
        array[j] = out[j];
}






void printArray(int array[], int size)
{
    printf("[");

    for (int i = 0; i < size; i++)
        printf("%d, ", array[i]);

    printf("\b\b]\n");
}







int main() 
{
    printf("\n\n ---------------- \n\n");
    printf("a simple test:\n\n");
    

    int array[] = {2, 12, 4, 0, 21, 23, 11, 43, 432, 130, 99, 33};

    printf("before sort: ");
    printArray(array, 12);


    counting_sort(array, 12, max(array, 12));


    printf("after sort: ");
    printArray(array, 12);


    printf("\n ---------------- \n\n\n");



    return 0;
}

File: Algorithms/counting sort/counting_sort.py








def counting_sort(array: list, n: int, k: int) -> list:

    count = [0] * (k+1)

    for j in range(n):
        count[array[j]] += 1

    for i in range(1, k+1):
        count[i] += count[i-1]

    out = [None] * n
    for j in range(n-1, -1, -1):
        out[count[array[j]]-1] = array[j]
        count[array[j]] -= 1


    return out






if __name__ == "__main__":

    print("\n\n ---------------- \n")
    print("a simple test:\n")
    

    array = [27, 21, 43, 50, 0, 2, 11, 77, 50, 101, 72, 16]
    print(f"before sort: {array}")

    array = counting_sort(array, len(array), 101)
    print(f"after sort: {array}")


    print("\n ---------------- \n\n")
